### ISSUE TITLE
nginx: intermittent / slow page loads when config'd in "server" block

### ISSUE THREAD

**Issue by eyz** [SpiderLabs/ModSecurity#855]
This is regarding the nginx version of ModSecurity 2.9.0 (master)

When enabling ModSecurity in a "location" block, performance is consistent and predictable.

When enabling ModSecurity in a "server" block, there are intermittent timeouts as seen from my testing under Chrome. Chrome would occasionally wait on the response, with only partial headers returned (if any). When the GET was successful (running in "server"), it generally would be received in about double the time that was experienced when running in "location".

I am using the following directives -

ModSecurityEnabled on;
ModSecurityConfig modsecurity.conf;
^ I can share the modsecurity.conf if that helps

I will gladly assist with any troubleshooting. I will close this issue if I find the issue no longer occurs upon subsequent builds.

nginx version: nginx/1.6.2 (on x86_64)
built by gcc 4.8.2 20140120 (Red Hat 4.8.2-16) (GCC) 
TLS SNI support enabled
configure arguments: --with-ld-opt=-Wl,-rpath,/usr/local/lib --with-cc-opt='-DNGX_LUA_USE_ASSERT -DNGX_LUA_ABORT_AT_PANIC' --add-module=/root/lua-nginx-module-0.9.16rc1 --add-module=/root/ngx_devel_kit-0.2.19 --add-module=/root/modsecurity-2.9.0/nginx/modsecurity --prefix=/usr/share/nginx --sbin-path=/usr/sbin/nginx --conf-path=/etc/nginx/nginx.conf --error-log-path=/var/log/nginx/error.log --http-log-path=/var/log/nginx/access.log --http-client-body-temp-path=/var/lib/nginx/tmp/client_body --http-proxy-temp-path=/var/lib/nginx/tmp/proxy --http-fastcgi-temp-path=/var/lib/nginx/tmp/fastcgi --http-uwsgi-temp-path=/var/lib/nginx/tmp/uwsgi --http-scgi-temp-path=/var/lib/nginx/tmp/scgi --pid-path=/var/run/nginx.pid --lock-path=/var/lock/subsys/nginx --user=nginx --group=nginx --with-file-aio --with-ipv6 --with-http_ssl_module --with-http_spdy_module --with-http_realip_module --with-http_addition_module --with-http_xslt_module --with-http_image_filter_module --with-http_geoip_module --with-http_sub_module --with-http_dav_module --with-http_flv_module --with-http_mp4_module --with-http_gunzip_module --with-http_gzip_static_module --with-http_random_index_module --with-http_secure_link_module --with-http_degradation_module --with-http_stub_status_module --with-http_perl_module --with-mail --with-mail_ssl_module --with-pcre --with-google_perftools_module --with-debug --with-cc-opt='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic' --with-ld-opt=-Wl,-E

**Comment by eksffa** [comment]
Confirmed, very same behavior for me on nginx 1.6.2 but 1.6.1 too and mod_security 2.9.0 but 2.8.0 as well. Noticed that enabling globally on http {} block has a predictable expected behavior again. The issue seems only on server{} block.


# uname -rms
FreeBSD 10.1-STABLE amd64

# nginx -V
nginx version: nginx/1.6.2
TLS SNI support enabled
configure arguments: --prefix=/usr/local/etc/nginx --with-cc-opt='-I /usr/local/include' --with-ld-opt='-L /usr/local/lib' --conf-path=/usr/local/etc/nginx/nginx.conf --sbin-path=/usr/local/sbin/nginx --pid-path=/var/run/nginx.pid --error-log-path=/var/log/nginx-error.log --user=www --group=www --with-ipv6 --http-client-body-temp-path=/var/tmp/nginx/client_body_temp --http-fastcgi-temp-path=/var/tmp/nginx/fastcgi_temp --http-proxy-temp-path=/var/tmp/nginx/proxy_temp --http-scgi-temp-path=/var/tmp/nginx/scgi_temp --http-uwsgi-temp-path=/var/tmp/nginx/uwsgi_temp --http-log-path=/var/log/nginx-access.log --add-module=/wrkdirs/usr/ports/proapps/nginx/work/giom-nginx_accept_language_module-2f69842 --add-module=/wrkdirs/usr/ports/proapps/nginx/work/nginx-accesskey-2.0.3 --with-http_addition_module --add-module=/wrkdirs/usr/ports/proapps/nginx/work/samizdatco-nginx-http-auth-digest-bd1c86a --add-module=/wrkdirs/usr/ports/proapps/nginx/work/http_auth_ldap-c4dc7c9153 --add-module=/wrkdirs/usr/ports/proapps/nginx/work/ngx_http_auth_request_module-0.2 --add-module=/wrkdirs/usr/ports/proapps/nginx/work/ngx_cache_purge-2.1 --add-module=/wrkdirs/usr/ports/proapps/nginx/work/openresty-echo-nginx-module-44c92a5 --add-module=/wrkdirs/usr/ports/proapps/nginx/work/openresty-headers-more-nginx-module-0c6e05d --add-module=/wrkdirs/usr/ports/proapps/nginx/work/vkholodkov-nginx-eval-module-125fa2e --add-module=/wrkdirs/usr/ports/proapps/nginx/work/ngx-fancyindex-0.3.1 --add-module=/wrkdirs/usr/ports/proapps/nginx/work/alibaba-nginx-http-footer-filter-3572184 --with-http_geoip_module --with-http_gzip_static_module --with-http_gunzip_module --with-http_image_filter_module --with-http_random_index_module --with-http_realip_module --add-module=/wrkdirs/usr/ports/proapps/nginx/work/ngx_http_response-0.3 --with-http_secure_link_module --with-http_stub_status_module --with-http_sub_module --add-module=/wrkdirs/usr/ports/proapps/nginx/work/mod_zip-1.1.6 --add-module=/wrkdirs/usr/ports/proapps/nginx/work/simpl-ngx_devel_kit-8dd0df5 --add-module=/wrkdirs/usr/ports/proapps/nginx/work/openresty-encrypted-session-nginx-module-49d741b --add-module=/wrkdirs/usr/ports/proapps/nginx/work/arut-nginx-let-module-a5e1dc5 --add-module=/wrkdirs/usr/ports/proapps/nginx/work/openresty-memc-nginx-module-de4cf86 --add-module=/wrkdirs/usr/ports/proapps/nginx/work/modsecurity-2.8.0/nginx/modsecurity --add-module=/wrkdirs/usr/ports/proapps/nginx/work/kyprizel-testcookie-nginx-module-f09f0db --add-module=/wrkdirs/usr/ports/proapps/nginx/work/naxsi-core-0.50/naxsi_src --with-pcre --add-module=/wrkdirs/usr/ports/proapps/nginx/work/openresty-rds-json-nginx-module-8292070 --add-module=/wrkdirs/usr/ports/proapps/nginx/work/openresty-xss-nginx-module-7e37038 --with-http_ssl_module --add-module=/wrkdirs/usr/ports/proapps/nginx/work/openresty-array-var-nginx-module-4676747

**Comment by uxbod** [comment]
Interestingly we are seeing very poor performance when in a location block using; and we are using the ngx_refactoring branch :

nginx -V
nginx version: nginx/1.9.0
built by gcc 4.8.2 20140120 (Red Hat 4.8.2-16) (GCC)
built with OpenSSL 1.0.1e-fips 11 Feb 2013
TLS SNI support enabled
configure arguments: --prefix=/etc/nginx --sbin-path=/usr/sbin/nginx --conf-path=/etc/nginx/nginx.conf --error-log-path=/var/log/nginx/error.log --http-log-path=/var/log/nginx/access.log --pid-path=/var/run/nginx.pid --lock-path=/var/run/nginx.lock --http-client-body-temp-path=/var/cache/nginx/client_temp --http-proxy-temp-path=/var/cache/nginx/proxy_temp --http-fastcgi-temp-path=/var/cache/nginx/fastcgi_temp --http-uwsgi-temp-path=/var/cache/nginx/uwsgi_temp --http-scgi-temp-path=/var/cache/nginx/scgi_temp --user=nginx --group=nginx --add-module=./naxsi/naxsi_src --add-module=./ngx_pagespeed --add-module=./modsecurity/nginx/modsecurity --add-module=./ngx_cache_purge --with-http_ssl_module --with-http_realip_module --with-http_addition_module --with-http_sub_module --with-http_dav_module --with-http_flv_module --with-http_mp4_module --with-http_gunzip_module --with-http_gzip_static_module --with-http_random_index_module --with-http_secure_link_module --with-http_stub_status_module --with-http_auth_request_module --with-mail --with-mail_ssl_module --with-file-aio --with-ipv6 --with-http_spdy_module --with-cc-opt='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic'

As soon as modsec is disabled then our Wordpress sites are lighting fast again.  The main area of concern is when in wp-admin as that becomes unusable.  We have tried using the following block:

location /wp-admin {
        index index.php;
        ModSecurityEnabled off;
        try_files $uri $uri/ /index.php$args;
}

but that does not help matters; though perhaps we have defined it incorrectly.

**Comment by ciarans** [comment]
+1 When enabled it is very slow

**Comment by zimmerle** [comment]
No longer a concern in libModSecurity. Marking it as won't fix for 2.x. Further information about libModSecurity available here:
https://github.com/SpiderLabs/ModSecurity/tree/v3/master
